lib/std/channels.nim

```diff
@@ -0,0 +1,697 @@
+##
+## The following is a simple example of two different ways to use channels:
+## blocking and non-blocking.
+## 
+
+runnableExamples("--threads:on --gc:orc"):
+  import std/os
+
+  # In this example a channel is declared at module scope.
+  # Channels are generic, and they include support for passing objects between
+  # threads.
+  # Note that isolated data passed through channels is moved around.
+  var chan = initChan[string]()
+
+  # This proc will be run in another thread using the threads module.
+  proc firstWorker() =
+    chan.send("Hello World!")
+
+  # This is another proc to run in a background thread. This proc takes a while
+  # to send the message since it sleeps for 2 seconds (or 2000 milliseconds).
+  proc secondWorker() =
+    sleep(2000)
+    chan.send("Another message")
+
+  # Launch the worker.
+  var worker1: Thread[void]
+  createThread(worker1, firstWorker)
+
+  # Block until the message arrives, then print it out.
+  var dest = ""
+  chan.recv(dest)
+  assert dest == "Hello World!"
+
+  # Wait for the thread to exit before moving on to the next example.
+  worker1.joinThread()
+
+  # Launch the other worker.
+  var worker2: Thread[void]
+  createThread(worker2, secondWorker)
+  # This time, use a non-blocking approach with tryRecv.
+  # Since the main thread is not blocked, it could be used to perform other
+  # useful work while it waits for data to arrive on the channel.
+  var messages: seq[string]
+  while true:
+    var msg = ""
+    if chan.tryRecv(msg):
+      messages.add msg # "Another message"
+      break
+
+    messages.add "Pretend I'm doing useful work..."
+    # For this example, sleep in order not to flood stdout with the above
+    # message.
+    sleep(400)
+
+  # Wait for the second thread to exit before cleaning up the channel.
+  worker2.joinThread()
+
+  # Clean up the channel.
+  assert chan.close()
+
+  assert messages[^1] == "Another message"
+  assert messages.len >= 2
+
+
+when not defined(gcArc) and not defined(gcOrc) and not defined(nimdoc):
+  {.error: "This channel implementation requires --gc:arc or --gc:orc".}
+
+import std/[locks, atomics, isolation]
+import system/ansi_c
+
+# Channel (Shared memory channels)
+# ----------------------------------------------------------------------------------
+
+const
+  CacheLineSize {.intdefine.} = 64 # TODO: some Samsung phone have 128 cache-line
+  ChannelCacheSize* {.intdefine.} = 100
+
+type
+  # ChannelBufKind = enum
+  #   Unbuffered # Unbuffered (blocking) channel
+  #   Buffered   # Buffered (non-blocking channel)
+
+  ChannelKind* = enum
+    Mpmc # Multiple producer, multiple consumer
+    Mpsc # Multiple producer, single consumer
+    Spsc # Single producer, single consumer
+
+  ChannelRaw* = ptr ChannelObj
+  ChannelObj = object
+    headLock, tailLock: Lock
+    notFullCond: Cond
+    notEmptyCond: Cond
+    owner: int32
+    impl: ChannelKind
+    closed: Atomic[bool]
+    size: int32
```

Comment:
Do we expect more than 2 billions items enqueued in a channel?

Reply:
hardly but `space: int` would be more idiomatic.

Reply:
I use `int`.